R-package/configure

```diff
@@ -1699,6 +1699,12 @@ CXX=`"${R_HOME}/bin/R" CMD config CXX11`
 # LightGBM-specific flags
 LGB_CPPFLAGS=""
 
+#########
```

Comment:
/gha run r-valgrind

Reply:
this build (https://github.com/microsoft/LightGBM/runs/1629232486?check_suite_focus=true) failed with an error I didn't see testing the CRAN package locally

> In file included from ./include/Eigen/Dense:1,
                 from treelearner/linear_tree_learner.cpp:7:
./include/Eigen/Core:15:10: fatal error: src/Core/util/DisableStupidWarnings.h: No such file or directory
   15 | #include "src/Core/util/DisableStupidWarnings.h"

Will look into it tomorrow.  Probably just a mistake I made with the include paths.

Reply:
/gha run r-valgrind

Reply:
https://github.com/microsoft/LightGBM/runs/1632880373?check_suite_focus=true

Reply:
/gha run r-valgrind

Reply:
https://github.com/microsoft/LightGBM/actions/runs/455828692

Reply:
the valgrind check is passing, but I think that's because the check is fragile. It seems that the linear-learner change did introduce some issues for valgrind.

```text
==2176== 
==2176== LEAK SUMMARY:
==2176==    definitely lost: 318 bytes in 10 blocks
==2176==    indirectly lost: 576 bytes in 9 blocks
==2176==      possibly lost: 336 bytes in 1 blocks
==2176==    still reachable: 312,088,759 bytes in 55,705 blocks
==2176==                       of which reachable via heuristic:
==2176==                         newarray           : 4,264 bytes in 1 blocks
==2176==         suppressed: 0 bytes in 0 blocks
==2176== Reachable blocks (those to which a pointer was found) are not shown.
==2176== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2176== 
==2176== For lists of detected and suppressed errors, rerun with: -s
==2176== ERROR SUMMARY: 153 errors from 14 contexts (suppressed: 0 from 0)
```

I'll open a separate issue 😞

Reply:
Ok I just tried on a PR into `master` and I'm convinced now that **the changes in #3299 introduced valgrind issues that would keep the R package off of CRAN**

This `valgrind` test against the current state of `master` (in which the R package does not allow using linear models at the leaves) are passing and the logs don't show anything that CRAN would consider problematic:

https://github.com/microsoft/LightGBM/runs/1634917288?check_suite_focus=true

```text
==1915== LEAK SUMMARY:
==1915==    definitely lost: 0 bytes in 0 blocks
==1915==    indirectly lost: 0 bytes in 0 blocks
==1915==      possibly lost: 336 bytes in 1 blocks
==1915==    still reachable: 308,145,480 bytes in 55,477 blocks
==1915==                       of which reachable via heuristic:
==1915==                         newarray           : 4,264 bytes in 1 blocks
==1915==         suppressed: 0 bytes in 0 blocks
==1915== Reachable blocks (those to which a pointer was found) are not shown.
==1915== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==1915== 
==1915== For lists of detected and suppressed errors, rerun with: -s
==1915== ERROR SUMMARY: 3 errors from 2 contexts (suppressed: 0 from 0)
```

So the issues that need to be fixed (from [this build](https://github.com/microsoft/LightGBM/runs/1632880373?check_suite_focus=true)) are as follows:

<details><summary>valgrind issues that need to be fixed (click me)</summary>

```text
Loading required package: R6
==2178== Conditional jump or move depends on uninitialised value(s)
==2178==    at 0x1541D563: LightGBM::LinearTreeLearner::Train(float const*, float const*, bool) (linear_tree_learner.cpp:110)
==2178==    by 0x1524FB12: LightGBM::GBDT::TrainOneIter(float const*, float const*) (gbdt.cpp:401)
==2178==    by 0x15585518: LightGBM::Booster::TrainOneIter() (c_api.cpp:351)
==2178==    by 0x155771F0: LGBM_BoosterUpdateOneIter (c_api.cpp:1637)
==2178==    by 0x155A5F13: LGBM_BoosterUpdateOneIter_R (lightgbm_R.cpp:399)
==2178==    by 0x4941E37: R_doDotCall (dotcode.c:601)
==2178==    by 0x494CFE9: do_dotcall (dotcode.c:1281)
==2178==    by 0x49A06EB: bcEval (eval.c:7072)
==2178==    by 0x498C2CC: Rf_eval (eval.c:727)
==2178==    by 0x498F03B: R_execClosure (eval.c:1897)
==2178==    by 0x498ECEE: Rf_applyClosure (eval.c:1823)
==2178==    by 0x498CAC7: Rf_eval (eval.c:850)
==2178==  Uninitialised value was created by a heap allocation
==2178==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2178==    by 0x1555CB8D: LightGBM::TreeLearner::CreateTreeLearner(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LightGBM::Config const*) (tree_learner.cpp:20)
==2178==    by 0x1524D381: LightGBM::GBDT::Init(LightGBM::Config const*, LightGBM::Dataset const*, LightGBM::ObjectiveFunction const*, std::vector<LightGBM::Metric const*, std::allocator<LightGBM::Metric const*> > const&) (gbdt.cpp:93)
==2178==    by 0x15582FCE: LightGBM::Booster::Booster(LightGBM::Dataset const*, char const*) (c_api.cpp:139)
==2178==    by 0x1557620A: LGBM_BoosterCreate (c_api.cpp:1535)
==2178==    by 0x155A54A5: LGBM_BoosterCreate_R (lightgbm_R.cpp:326)
==2178==    by 0x4941EAF: R_doDotCall (dotcode.c:607)
==2178==    by 0x494CFE9: do_dotcall (dotcode.c:1281)
==2178==    by 0x49A06EB: bcEval (eval.c:7072)
==2178==    by 0x498C2CC: Rf_eval (eval.c:727)
==2178==    by 0x498F03B: R_execClosure (eval.c:1897)
==2178==    by 0x498ECEE: Rf_applyClosure (eval.c:1823)
==2178== 
==2178== Conditional jump or move depends on uninitialised value(s)
==2178==    at 0x1541E80A: LightGBM::LinearTreeLearner::AddPredictionToScore(LightGBM::Tree const*, double*) const (linear_tree_learner.h:44)
==2178==    by 0x1525561F: LightGBM::ScoreUpdater::AddScore(LightGBM::TreeLearner const*, LightGBM::Tree const*, int) (score_updater.hpp:91)
==2178==    by 0x15250503: LightGBM::GBDT::UpdateScore(LightGBM::Tree const*, int) (gbdt.cpp:495)
==2178==    by 0x1524FCB7: LightGBM::GBDT::TrainOneIter(float const*, float const*) (gbdt.cpp:413)
==2178==    by 0x15585518: LightGBM::Booster::TrainOneIter() (c_api.cpp:351)
==2178==    by 0x155771F0: LGBM_BoosterUpdateOneIter (c_api.cpp:1637)
==2178==    by 0x155A5F13: LGBM_BoosterUpdateOneIter_R (lightgbm_R.cpp:399)
==2178==    by 0x4941E37: R_doDotCall (dotcode.c:601)
==2178==    by 0x494CFE9: do_dotcall (dotcode.c:1281)
==2178==    by 0x49A06EB: bcEval (eval.c:7072)
==2178==    by 0x498C2CC: Rf_eval (eval.c:727)
==2178==    by 0x498F03B: R_execClosure (eval.c:1897)
==2178==  Uninitialised value was created by a heap allocation
==2178==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2178==    by 0x1555CB8D: LightGBM::TreeLearner::CreateTreeLearner(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LightGBM::Config const*) (tree_learner.cpp:20)
==2178==    by 0x1524D381: LightGBM::GBDT::Init(LightGBM::Config const*, LightGBM::Dataset const*, LightGBM::ObjectiveFunction const*, std::vector<LightGBM::Metric const*, std::allocator<LightGBM::Metric const*> > const&) (gbdt.cpp:93)
==2178==    by 0x15582FCE: LightGBM::Booster::Booster(LightGBM::Dataset const*, char const*) (c_api.cpp:139)
==2178==    by 0x1557620A: LGBM_BoosterCreate (c_api.cpp:1535)
==2178==    by 0x155A54A5: LGBM_BoosterCreate_R (lightgbm_R.cpp:326)
==2178==    by 0x4941EAF: R_doDotCall (dotcode.c:607)
==2178==    by 0x494CFE9: do_dotcall (dotcode.c:1281)
==2178==    by 0x49A06EB: bcEval (eval.c:7072)
==2178==    by 0x498C2CC: Rf_eval (eval.c:727)
==2178==    by 0x498F03B: R_execClosure (eval.c:1897)
==2178==    by 0x498ECEE: Rf_applyClosure (eval.c:1823)
==2178== 19 bytes in 1 blocks are definitely lost in loss record 31 of 2,742
==2178==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2178==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2178==    by 0x155672FC: LightGBM::Config::Config() (config.h:34)
==2178==    by 0x15575DC8: LGBM_DatasetUpdateParamChecking (c_api.cpp:1496)
==2178==    by 0x155A4F52: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2178==    by 0x4941E6F: R_doDotCall (dotcode.c:604)
==2178==    by 0x494CFE9: do_dotcall (dotcode.c:1281)
==2178==    by 0x498C97C: Rf_eval (eval.c:830)
==2178==    by 0x4992CE6: do_set (eval.c:2969)
==2178==    by 0x498C723: Rf_eval (eval.c:802)
==2178==    by 0x4991769: do_begin (eval.c:2517)
==2178==    by 0x498C723: Rf_eval (eval.c:802)
==2178== 
==2178== 19 bytes in 1 blocks are definitely lost in loss record 32 of 2,742
==2178==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2178==    by 0x15245A23: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_M_allocate(unsigned long) (stl_vector.h:346)
==2178==    by 0x152429BC: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_M_create_storage(unsigned long) (stl_vector.h:361)
==2178==    by 0x1523E066: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_Vector_base(unsigned long, std::allocator<std::vector<double, std::allocator<double> > > const&) (stl_vector.h:305)
==2178==    by 0x152FF682: std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::vector(unsigned long, std::vector<double, std::allocator<double> > const&, std::allocator<std::vector<double, std::allocator<double> > > const&) (stl_vector.h:524)
==2178==    by 0x152F94D7: LightGBM::Config::GetAucMuWeights() (config.cpp:159)
==2178==    by 0x152F9CE2: LightGBM::Config::Set(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (config.cpp:219)
==2178==    by 0x1558377D: LightGBM::Booster::CheckDatasetResetConfig(LightGBM::Config const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (c_api.cpp:190)
==2178==    by 0x15575E13: LGBM_DatasetUpdateParamChecking (c_api.cpp:1499)
==2178==    by 0x155A4F52: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2178== 
==2178== 336 bytes in 1 blocks are possibly lost in loss record 161 of 2,742
==2178==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2178==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==2178==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==2178==    by 0x5706322: allocate_stack (allocatestack.c:622)
==2178==    by 0x5706322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==2178==    by 0x56D4DDA: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2178==    by 0x56CC8E0: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2178==    by 0x15574879: LGBM_DatasetCreateFromCSC (c_api.cpp:1303)
==2178==    by 0x155A36CC: LGBM_DatasetCreateFromCSC_R (lightgbm_R.cpp:91)
==2178==    by 0x4942152: R_doDotCall (dotcode.c:634)
==2178==    by 0x494CFE9: do_dotcall (dotcode.c:1281)
==2178==    by 0x49A06EB: bcEval (eval.c:7072)
==2178==    by 0x498C2CC: Rf_eval (eval.c:727)
==2178==    by 0x498F03B: R_execClosure (eval.c:1897)
==2178== 
==2178== 664 (104 direct, 560 indirect) bytes in 1 blocks are definitely lost in loss record 210 of 2,742
==2178==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2178==    by 0x1526815F: __gnu_cxx::new_allocator<std::__detail::_Hash_node_base*>::allocate(unsigned long, void const*) (new_allocator.h:115)
==2178==    by 0x15267970: std::allocator_traits<std::allocator<std::__detail::_Hash_node_base*> >::allocate(std::allocator<std::__detail::_Hash_node_base*>&, unsigned long) (alloc_traits.h:460)
==2178==    by 0x1527CDF8: std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true> > >::_M_allocate_buckets(unsigned long) (hashtable_policy.h:2085)
==2178==    by 0x1527C10E: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_allocate_buckets(unsigned long) (hashtable.h:398)
==2178==    by 0x1527ABB0: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_rehash_aux(unsigned long, std::integral_constant<bool, true>) (hashtable.h:2091)
==2178==    by 0x1527932F: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_rehash(unsigned long, unsigned long const&) (hashtable.h:2070)
==2178==    by 0x15277AD1: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_insert_unique_node(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, unsigned long, std::__detail::_Hash_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>*, unsigned long) (hashtable.h:1728)
==2178==    by 0x15300CCD: std::pair<std::__detail::_Node_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false, true>, bool> std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::integral_constant<bool, true>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (hashtable.h:1682)
==2178==    by 0x152FFD76: std::pair<std::__detail::_Node_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false, true>, bool> std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (hashtable.h:788)
==2178==    by 0x152FF4E3: std::pair<std::__detail::_Node_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false, true>, bool> std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (unordered_map.h:388)
==2178==    by 0x152F7331: LightGBM::Config::KV2Map(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >*, char const*) (config.cpp:27)
```

</details>

I think this is outside of my C++ abilities. @guolinke or @btrotta are you able to help?

Reply:
I'll take a look

Reply:
@jameslamb I think I've fixed the first of the issues (see PR #3709). I'm confused about what's causing the others though, they don't seem to be related to the new code added for linear trees. The issue in `GetAucMuWeights` is strange, it seems to be complaining about the part where it allocates a vector of vectors, but I don't see why this would cause a leak. @guolinke let me know if you have any ideas.

Reply:
Thanks!

Reply:
/gha run r-valgrind

Reply:
let's see if #3709 helped:

https://github.com/microsoft/LightGBM/runs/1681914835?check_suite_focus=true

Reply:
It looks like #3709 did help with valgrind, thanks! I now see this

```text
==2174== 
==2174== LEAK SUMMARY:
==2174==    definitely lost: 318 bytes in 10 blocks
==2174==    indirectly lost: 576 bytes in 9 blocks
==2174==      possibly lost: 336 bytes in 1 blocks
==2174==    still reachable: 312,467,255 bytes in 55,720 blocks
==2174==                       of which reachable via heuristic:
==2174==                         newarray           : 4,264 bytes in 1 blocks
==2174==         suppressed: 0 bytes in 0 blocks
==2174== Reachable blocks (those to which a pointer was found) are not shown.
==2174== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2174== 
==2174== For lists of detected and suppressed errors, rerun with: -s
==2174== ERROR SUMMARY: 13 errors from 12 contexts (suppressed: 0 from 0)
```

So there are still issues that have to be addressed to stay in compliance with CRAN, but #3709 reduced the errors found from 153 to 14.

The other issues seem to mostly come from `LGBM_DatasetUpdateParamChecking`.

<details><summary>remaining issues to be fixed</summary>

```text
==2174== 
==2174== HEAP SUMMARY:
==2174==     in use at exit: 312,468,485 bytes in 55,740 blocks
==2174==   total heap usage: 2,839,931 allocs, 2,784,191 frees, 6,180,317,716 bytes allocated
==2174== 
==2174== 19 bytes in 1 blocks are definitely lost in loss record 30 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2174==    by 0x15568E06: LightGBM::Config::Config() (config.h:34)
==2174==    by 0x155778D2: LGBM_DatasetUpdateParamChecking (c_api.cpp:1498)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x498C970: Rf_eval (eval.c:830)
==2174==    by 0x4992CDA: do_set (eval.c:2969)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174==    by 0x499175D: do_begin (eval.c:2517)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174== 
==2174== 19 bytes in 1 blocks are definitely lost in loss record 31 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2174==    by 0x15568E06: LightGBM::Config::Config() (config.h:34)
==2174==    by 0x1558526E: LightGBM::Booster::CheckDatasetResetConfig(LightGBM::Config const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (c_api.cpp:189)
==2174==    by 0x1557791D: LGBM_DatasetUpdateParamChecking (c_api.cpp:1501)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x498C970: Rf_eval (eval.c:830)
==2174==    by 0x4992CDA: do_set (eval.c:2969)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174==    by 0x499175D: do_begin (eval.c:2517)
==2174== 
==2174== 20 bytes in 1 blocks are definitely lost in loss record 32 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2174==    by 0x15569145: LightGBM::Config::Config() (config.h:34)
==2174==    by 0x155778D2: LGBM_DatasetUpdateParamChecking (c_api.cpp:1498)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x498C970: Rf_eval (eval.c:830)
==2174==    by 0x4992CDA: do_set (eval.c:2969)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174==    by 0x499175D: do_begin (eval.c:2517)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174== 
==2174== 20 bytes in 1 blocks are definitely lost in loss record 33 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2174==    by 0x15569145: LightGBM::Config::Config() (config.h:34)
==2174==    by 0x1558526E: LightGBM::Booster::CheckDatasetResetConfig(LightGBM::Config const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (c_api.cpp:189)
==2174==    by 0x1557791D: LGBM_DatasetUpdateParamChecking (c_api.cpp:1501)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x498C970: Rf_eval (eval.c:830)
==2174==    by 0x4992CDA: do_set (eval.c:2969)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174==    by 0x499175D: do_begin (eval.c:2517)
==2174== 
==2174== 28 bytes in 1 blocks are definitely lost in loss record 49 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2174==    by 0x155690D3: LightGBM::Config::Config() (config.h:34)
==2174==    by 0x155778D2: LGBM_DatasetUpdateParamChecking (c_api.cpp:1498)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x498C970: Rf_eval (eval.c:830)
==2174==    by 0x4992CDA: do_set (eval.c:2969)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174==    by 0x499175D: do_begin (eval.c:2517)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174== 
==2174== 28 bytes in 1 blocks are definitely lost in loss record 50 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x735786C: void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28)
==2174==    by 0x155690D3: LightGBM::Config::Config() (config.h:34)
==2174==    by 0x1558526E: LightGBM::Booster::CheckDatasetResetConfig(LightGBM::Config const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (c_api.cpp:189)
==2174==    by 0x1557791D: LGBM_DatasetUpdateParamChecking (c_api.cpp:1501)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x498C970: Rf_eval (eval.c:830)
==2174==    by 0x4992CDA: do_set (eval.c:2969)
==2174==    by 0x498C717: Rf_eval (eval.c:802)
==2174==    by 0x499175D: do_begin (eval.c:2517)
==2174== 
==2174== 32 (24 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 54 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x1524A2A8: __gnu_cxx::new_allocator<std::vector<double, std::allocator<double> > >::allocate(unsigned long, void const*) (new_allocator.h:115)
==2174==    by 0x15248BC2: std::allocator_traits<std::allocator<std::vector<double, std::allocator<double> > > >::allocate(std::allocator<std::vector<double, std::allocator<double> > >&, unsigned long) (alloc_traits.h:460)
==2174==    by 0x15246AE3: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_M_allocate(unsigned long) (stl_vector.h:346)
==2174==    by 0x15243A7C: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_M_create_storage(unsigned long) (stl_vector.h:361)
==2174==    by 0x1523F126: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_Vector_base(unsigned long, std::allocator<std::vector<double, std::allocator<double> > > const&) (stl_vector.h:305)
==2174==    by 0x15300A28: std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::vector(unsigned long, std::vector<double, std::allocator<double> > const&, std::allocator<std::vector<double, std::allocator<double> > > const&) (stl_vector.h:524)
==2174==    by 0x152FA87D: LightGBM::Config::GetAucMuWeights() (config.cpp:159)
==2174==    by 0x152FB088: LightGBM::Config::Set(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (config.cpp:219)
==2174==    by 0x155778EB: LGBM_DatasetUpdateParamChecking (c_api.cpp:1499)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174==    by 0x4941E63: R_doDotCall (dotcode.c:604)
==2174== 
==2174== 32 bytes in 1 blocks are definitely lost in loss record 55 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x15249BAB: __gnu_cxx::new_allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::allocate(unsigned long, void const*) (new_allocator.h:115)
==2174==    by 0x15248177: std::allocator_traits<std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::allocate(std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&, unsigned long) (alloc_traits.h:460)
==2174==    by 0x15245C25: std::_Vector_base<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::_M_allocate(unsigned long) (stl_vector.h:346)
==2174==    by 0x153013B2: void std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::_M_realloc_insert<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>(__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (vector.tcc:440)
==2174==    by 0x153009B1: std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::push_back(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (stl_vector.h:1198)
==2174==    by 0x152F91C6: LightGBM::ParseMetrics(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) (config.cpp:74)
==2174==    by 0x152F94C7: LightGBM::GetMetricType(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) (config.cpp:92)
==2174==    by 0x152FAFB6: LightGBM::Config::Set(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (config.cpp:209)
==2174==    by 0x15585287: LightGBM::Booster::CheckDatasetResetConfig(LightGBM::Config const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (c_api.cpp:190)
==2174==    by 0x1557791D: LGBM_DatasetUpdateParamChecking (c_api.cpp:1501)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174== 
==2174== 32 (24 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 56 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x1524A2A8: __gnu_cxx::new_allocator<std::vector<double, std::allocator<double> > >::allocate(unsigned long, void const*) (new_allocator.h:115)
==2174==    by 0x15248BC2: std::allocator_traits<std::allocator<std::vector<double, std::allocator<double> > > >::allocate(std::allocator<std::vector<double, std::allocator<double> > >&, unsigned long) (alloc_traits.h:460)
==2174==    by 0x15246AE3: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_M_allocate(unsigned long) (stl_vector.h:346)
==2174==    by 0x15243A7C: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_M_create_storage(unsigned long) (stl_vector.h:361)
==2174==    by 0x1523F126: std::_Vector_base<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::_Vector_base(unsigned long, std::allocator<std::vector<double, std::allocator<double> > > const&) (stl_vector.h:305)
==2174==    by 0x15300A28: std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::vector(unsigned long, std::vector<double, std::allocator<double> > const&, std::allocator<std::vector<double, std::allocator<double> > > const&) (stl_vector.h:524)
==2174==    by 0x152FA87D: LightGBM::Config::GetAucMuWeights() (config.cpp:159)
==2174==    by 0x152FB088: LightGBM::Config::Set(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (config.cpp:219)
==2174==    by 0x15585287: LightGBM::Booster::CheckDatasetResetConfig(LightGBM::Config const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) (c_api.cpp:190)
==2174==    by 0x1557791D: LGBM_DatasetUpdateParamChecking (c_api.cpp:1501)
==2174==    by 0x155A6D98: LGBM_DatasetUpdateParamChecking_R (lightgbm_R.cpp:285)
==2174== 
==2174== 336 bytes in 1 blocks are possibly lost in loss record 163 of 2,743
==2174==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==2174==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==2174==    by 0x5706322: allocate_stack (allocatestack.c:622)
==2174==    by 0x5706322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==2174==    by 0x56D4DDA: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2174==    by 0x56CC8E0: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2174==    by 0x15576383: LGBM_DatasetCreateFromCSC (c_api.cpp:1305)
==2174==    by 0x155A5512: LGBM_DatasetCreateFromCSC_R (lightgbm_R.cpp:91)
==2174==    by 0x4942146: R_doDotCall (dotcode.c:634)
==2174==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2174==    by 0x49A06DF: bcEval (eval.c:7072)
==2174==    by 0x498C2C0: Rf_eval (eval.c:727)
==2174==    by 0x498F02F: R_execClosure (eval.c:1897)
==2174== 
==2174== 664 (104 direct, 560 indirect) bytes in 1 blocks are definitely lost in loss record 211 of 2,743
==2174==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2174==    by 0x1526921F: __gnu_cxx::new_allocator<std::__detail::_Hash_node_base*>::allocate(unsigned long, void const*) (new_allocator.h:115)
==2174==    by 0x15268A30: std::allocator_traits<std::allocator<std::__detail::_Hash_node_base*> >::allocate(std::allocator<std::__detail::_Hash_node_base*>&, unsigned long) (alloc_traits.h:460)
==2174==    by 0x1527E18C: std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true> > >::_M_allocate_buckets(unsigned long) (hashtable_policy.h:2085)
==2174==    by 0x1527D484: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_allocate_buckets(unsigned long) (hashtable.h:398)
==2174==    by 0x1527BED8: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_rehash_aux(unsigned long, std::integral_constant<bool, true>) (hashtable.h:2091)
==2174==    by 0x1527A5CD: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_rehash(unsigned long, unsigned long const&) (hashtable.h:2070)
==2174==    by 0x15278CF3: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_insert_unique_node(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, unsigned long, std::__detail::_Hash_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>*, unsigned long) (hashtable.h:1728)
==2174==    by 0x15302073: std::pair<std::__detail::_Node_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false, true>, bool> std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::integral_constant<bool, true>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (hashtable.h:1682)
==2174==    by 0x1530111C: std::pair<std::__detail::_Node_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false, true>, bool> std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (hashtable.h:788)
==2174==    by 0x15300889: std::pair<std::__detail::_Node_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, false, true>, bool> std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (unordered_map.h:388)
==2174==    by 0x152F86D7: LightGBM::Config::KV2Map(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >*, char const*) (config.cpp:27)
==2174== 
==2174== LEAK SUMMARY:
==2174==    definitely lost: 318 bytes in 10 blocks
==2174==    indirectly lost: 576 bytes in 9 blocks
==2174==      possibly lost: 336 bytes in 1 blocks
==2174==    still reachable: 312,467,255 bytes in 55,720 blocks
==2174==                       of which reachable via heuristic:
==2174==                         newarray           : 4,264 bytes in 1 blocks
==2174==         suppressed: 0 bytes in 0 blocks
==2174== Reachable blocks (those to which a pointer was found) are not shown.
==2174== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2174== 
==2174== For lists of detected and suppressed errors, rerun with: -s
==2174== ERROR SUMMARY: 13 errors from 12 contexts (suppressed: 0 from 0)
writing valgrind output to valgrind-logs.log
valgrind found 318 bytes definitely lost
```

</details>

Reply:
I looked into this some more, and it seems some of the errors occur when we attempt to reset the config and change a property that can't be changed after constructing the dataset. In this PR this is caused by the test where we attempt to change `linear_tree`. But I think the problem is not related to linear trees specifically. I experimented by running valgrind with just the one test below (which attempts to modify `max_bin` after constructing the dataset), and am getting similar errors. 

```
test_that("lgb.train() fails already-constructed dataset when changing max_bin", {
  set.seed(708L)
  params <- list(
    objective = "regression"
    , verbose = -1L
    , metric = "mse"
    , seed = 0L
    , num_leaves = 2L
  )

  dtrain <- lgb.Dataset(
    data = matrix(rnorm(100L), ncol = 1L)
    , label = rnorm(100L)
  )
  dtrain$construct()
  expect_error({
    bst_linear <- lgb.train(
      data = dtrain
      , nrounds = 10L
      , params = modifyList(params, list(max_bin = 20))
      , valids = list("train" = dtrain)
    )
  }, regexp = "Cannot change max_bin after constructed Dataset handle")
})
```
Log attached:
[one_test_only.log](https://github.com/microsoft/LightGBM/files/5801340/one_test_only.log)

Reply:
Maybe the problem is that the fatal error causes lightgbm to exit without freeing some allocated memory? Similar to this: https://stackoverflow.com/questions/1901322/valgrind-reports-memory-leak-when-assigning-a-value-to-a-string

Reply:
good idea! I know we ran into similar issues in the past. I'm going to try pushing a commit right now that skips those tests, to test that theory.

Reply:
/gha run r-valgrind

Reply:
https://github.com/microsoft/LightGBM/actions/runs/481344404

Reply:
nice work @btrotta ! After skipping the one test introduced in this PR that checks for errors, `valgrind` is happy :grinning: 

```text
==2177== 
==2177== HEAP SUMMARY:
==2177==     in use at exit: 312,459,948 bytes in 55,720 blocks
==2177==   total heap usage: 2,837,681 allocs, 2,781,961 frees, 6,180,339,569 bytes allocated
==2177== 
==2177== 336 bytes in 1 blocks are possibly lost in loss record 152 of 2,721
==2177==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2177==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==2177==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==2177==    by 0x5706322: allocate_stack (allocatestack.c:622)
==2177==    by 0x5706322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==2177==    by 0x56D4DDA: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2177==    by 0x56CC8E0: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==2177==    by 0x15576383: LGBM_DatasetCreateFromCSC (c_api.cpp:1305)
==2177==    by 0x155A5512: LGBM_DatasetCreateFromCSC_R (lightgbm_R.cpp:91)
==2177==    by 0x4942146: R_doDotCall (dotcode.c:634)
==2177==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==2177==    by 0x49A06DF: bcEval (eval.c:7072)
==2177==    by 0x498C2C0: Rf_eval (eval.c:727)
==2177==    by 0x498F02F: R_execClosure (eval.c:1897)
==2177== 
==2177== LEAK SUMMARY:
==2177==    definitely lost: 0 bytes in 0 blocks
==2177==    indirectly lost: 0 bytes in 0 blocks
==2177==      possibly lost: 336 bytes in 1 blocks
==2177==    still reachable: 312,459,612 bytes in 55,719 blocks
==2177==                       of which reachable via heuristic:
==2177==                         newarray           : 4,264 bytes in 1 blocks
==2177==         suppressed: 0 bytes in 0 blocks
==2177== Reachable blocks (those to which a pointer was found) are not shown.
==2177== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2177== 
==2177== For lists of detected and suppressed errors, rerun with: -s
==2177== ERROR SUMMARY: 3 errors from 2 contexts (suppressed: 0 from 0)
writing valgrind output to valgrind-logs.log
valgrind found 0 bytes definitely lost
valgrind found 0 bytes indirectly lost
valgrind found 336 bytes possibly lost
```

I think we should leave this skip in to get this feature merged, and I can work on removing it separately as a follow-up task.